.config-schema.json

```diff
@@ -419,6 +419,47 @@
                     "x-env-variable": "OPENFGA_CHECK_QUERY_CACHE_TTL"
                 }
             }
+        },
+        "dispatchThrottling": {
+            "type": "object",
+            "properties": {
+                "enabled": {
+                    "description": "enable throttling when request's number of dispatches is high",
+                    "type": "bool",
+                    "default": "false",
+                    "x-env-variable": "OPENFGA_DISPATCH_THROTTLING_ENABLED"
+                },
+                "timeTickerFrequency": {
+                    "description": "defines how frequent dispatch throttling will be evaluated",
+                    "type": "duration",
+                    "default": "10µs",
+                    "x-env-variable": "OPENFGA_DISPATCH_THROTTLING_TIME_TICKER_FREQUENCY"
+                },
+                "lowPriorityLevel": {
+                    "description": "define the number of dispatches to be considered in low priority throttling queue",
+                    "type": "integer",
+                    "default": "100",
+                    "x-env-variable": "OPENFGA_DISPATCH_THROTTLING_LOW_PRIORITY_LEVEL"
+                },
+                "lowPriorityShaper": {
+                    "description": "number of tickers required to dispatch a low priority work",
+                    "type": "integer",
+                    "default": "30",
+                    "x-env-variable": "OPENFGA_DISPATCH_THROTTLING_LOW_PRIORITY_SHAPER"
+                },
+                "mediumPriorityLevel": {
+                    "description": "define the number of dispatches to be considered in medium priority throttling queue",
+                    "type": "integer",
+                    "default": "50",
+                    "x-env-variable": "OPENFGA_DISPATCH_THROTTLING_MEDIUM_PRIORITY_LEVEL"
+                },
+                "mediumPriorityShaper": {
+                    "description": "initial frequency on un-throttled dispatches for medium priority jobs",
+                    "type": "integer",
+                    "default": "8",
+                    "x-env-variable": "OPENFGA_DISPATCH_THROTTLING_MEDIUM_PRIORITY_SHAPER"
+                }
+            }
```

Comment:
Personally I feel that such minute way to configure the system is a good thing. If we consider tools like RDBMS (postgres) or file systems, they have such fine tuned way to configure things as the ones who will be tweaking these are engineers and they might want to change these values for their specific edge cases.

Reply:
> could we not have a singular throttling threshold that limits this huge disparity.

I agree with this though, instead of having 2 queues. Why not have a single queue? Having a single DispatchThrottledQueue makes it much simpler than a LowPriorityQueue and a MediumPriorityQueue

About the requests not being gradual, why not we a single "THRESHOLD_VALUE" after which we send the request again to the back of the queue

Reply:
Even if we have a single "threshold_value", we will still need three knobs 
- level where we start shaping
- how frequently we shape at the beginning (we need to do that, otherwise, there is a sharp rise in latency between requests just over the  threshold vs. requests just below the threshold
- this single "threshold value" where we send the request again to the back of the queue

One concern with the single queue model is that the medium dispatches  (not too bad) are competing against the large dispatches.  We had seen in some cases where some model yield number of dispatches > 1000.  In this case, if we want to start shaping at 50 dispatches, requests with 60 dispatches will be competing against requests with 900 dispatches. This will make it unfair even when you send the large requests back to the line.  An analogy is that you have a bus with two groups (one group with 10 folks and another group with 1000 folks) and the folks line up randomly.  If you have a single line but penalize group 2 members by having them to go back to the line again, group 1's member will still have a hard time getting in the bus by virtue they have to compete against group 2.  If you have two lines, however, group 1 will have a better chance of getting into the bus because they are only competing against yourself.

For operators who only need a single queue, they can set the medium threshold to be equal low threshold.  In this way, they will effectively have a single queue. Perhaps I can update the configuration to point out that if only one queue is desired, set medium threshold level to be equal to low threshold level.

Reply:
@miparnisari @jpadilla I'm curious what you two also think of this expansion of configuration and the implementation overhead?

As a reminder for us all reading this thread, the original premise and goal was simple - **throttle requests that reach a dispatch count more than some threshold, because after some threshold we cannot guarantee our SLO will be met**. 

A single dispatch leads to at least 1 database query, but our SLO targets 50ms. Subproblems of a dispatch are evaluated concurrently with some maximum bound of concurrency. So I leave a question to the reader - On average, how many dispatches does it take to exceed a 50ms budget? If we don't have an answer to that, then we should evaluate that. This metric will obviously depend on the model and the tuple data. 

My hot take on these changes - I'm very hesitant to introduce this level of configuration to the server without comparing the alternative implementation which was proposed originally prior to this work kicking off.  Namely, throttle proportionally to some configured threshold using an exponentially delaying ticker. Such an implementation would be much easier to maintain as a team and easier to configure and to explain to the community how to configure. I recognize this kind of implementation _could_ have some implications, but both implementations have implications. It's our job to compare the implications relative to one another, and I don't think we have the information needed to make that determination yet.

I wouldn't know how to tune these existing settings, and I also wouldn't even know where to start to explain how to configure these settings to an external OpenFGA operator. These default values were based on a distribution that is unique to Okta FGA traffic, which isn't necessarily representative of the average case. Would these defaults work well for the average case?


One last thing - a lot of these configs are oriented around trying to manage the "shape" of the distribution, but instead of observing the distribution and apply the throttle we are trying to approximate the shape with synthentics represented as `OPENFGA_DISPATCH_THROTTLING_LOW_PRIORITY_SHAPER` and `OPENFGA_DISPATCH_THROTTLING_MEDIUM_PRIORITY_SHAPER`. If we're going to do multi-priority throttling, then I may suggest we compute the quantiles of the dispatches in real-time using an online approximation of them (t-digest or the like) and then assign requests whose dispatch count is greater than p99 to the low priority and any others >threshold but <= p99 to the high priority queue. This would achieve the shaping we're seeking and would only require 2 queues and would also avoid the overhead of all of these configs.

Reply:
Updated to use only 1 level.  We can add the multi-level back when it is needed.